lib/spack/spack/installer.py

```diff
@@ -932,6 +934,7 @@ def _ensure_locked(self, lock_type, pkg):
         except (lk.LockDowngradeError, lk.LockTimeoutError) as exc:
             tty.debug(err.format(op, desc, pkg_id, exc.__class__.__name__,
```

Comment:
> (main question) In the case of a `LockDowngradeError` I think that implies a bug in Spack - in that case it should probably terminate. Is it clear why this occurs and/or what the read/write counts are at the time of failure?

It was my impression that the read/write counts are intended primarily for tracking nested locking for a process.  In the installation context for a spec, they are generally 0 or 1.  In the context of an environment installation with a broken package that assumption appears to break down.  

I don't have notes on exactly what the counts were but I do recall that they increased with each iteration on a package that had a cached lock in place.

The problem with terminating on that failure is we don't then perform a 'best effort' installation in the environment context, which is the whole point of this PR.

In particular, IIRC the (broken) `slate` package within the aforementioned environment was causing a `SystemExit` that precluded lock cleanup. So when a package 

According to my notes, adding the `try-except` block for the environment's call to `do_install` exposed the downgrade lock issue.  I didn't keep notes on the exact prefix lock read/write counts stored in the database cache, but I do recall they were not as expected.

Reply:
> In particular, IIRC the (broken) slate package within the aforementioned environment was causing a SystemExit that precluded lock cleanup. So when a package

I haven't parsed this fully yet but this line appears incomplete

Reply:
> > In particular, IIRC the (broken) slate package within the aforementioned environment was causing a SystemExit that precluded lock cleanup. So when a package
> 
> I haven't parsed this fully yet but this line appears incomplete

Thanks for catching that.  That might have been a result of some of the "help" one of the foster kittens was providing when he walked on the laptop keyboard.

Reply:
If the locks are reentrant and allow for counts that are above 1. Then perhaps the logic in `llnl.util.lock` should be updated. For example

```
    def downgrade_write_to_read(self, timeout=None):
 
        if self._writes == 1 and self._reads == 0:
```

could change to 

```
    def downgrade_write_to_read(self, timeout=None):
 
        if self._writes == 1:  # no need to count how many read locks there are
```

?

Reply:
> If the locks are reentrant and allow for counts that are above 1. Then perhaps the logic in `llnl.util.lock` should be updated. For example
> 
> ```
>     def downgrade_write_to_read(self, timeout=None):
>  
>         if self._writes == 1 and self._reads == 0:
> ```
> 
> could change to
> 
> ```
>     def downgrade_write_to_read(self, timeout=None):
>  
>         if self._writes == 1:  # no need to count how many read locks there are
> ```
> 
> ?

Perhaps.  I was precluding nested transactions for the downgrade.  I'd appreciate @tgamblin 's perspective on this since he's the locking expert.